### DISCUSSION TITLE
subscriptions - optimistic updates

### DISCUSSION THREAD

**Discussion by jdgamble555** [urql-graphql/urql#1423] [Q&A] [answered]
Is it possible to do subscriptions with optimistic updates?

This is necessary as the DOM does not get updated until the subscription returns data back.  Updating the DOM first, optimistically, would result in a fast, immediate, responsive app.

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

This sound a bit counter-intuitive, a subscription isn't meant to have a predictable result, it's more that you are subscribed to a certain entity like a chat that can have updates at any given time, this way you can stay up-to-date with you updating stuff in other tabs or another user sending a message.

What is your use case specifically for "optimistic responses" in subscriptions.

I understand the case for mutations where we know what we are changing, for instance we change a todo.title to be value y and we can change that in the cache before the server responds.

**Reply by jdgamble555** [reply]
That is the use case.  I have subscriptions displayed on the page, I update a mutation (knowing the result), and it gets updated in the DOM immediately, instead of waiting up to a few seconds for the result to come back.  This is how **firebase** works, the change is immediate in the DOM.

I know this is not possible in [Apollo](https://github.com/apollographql/apollo-client/issues/5267#issue-488242271) directly, but I definitely see a use case for it.  You can see I was able to emulate what I am talking about on the front end [here](https://github.com/jdgamble555/angular-fire-dgraph/blob/master/src/app/core/dgraph.service.ts#L72) by updating the frontend first before the subscription gets updated.  I would see many advantages to this updating as it in the cache, as it would give you a faster reactive app.

**Reply by JoviDeCroock** [reply]
I mean you can optimistically update from the outgoing mutation. It's impossible to predict what a subscription will return, that's kind off against the pattern.

**Reply by jdgamble555** [reply]
Yes, I am aware of that.  The problem is that **mutations** are updated in the cache as well as **queries**.  However, you cannot run **optimistic updates** on **subscriptions** because it does not save in the **cache**.  

I am proposing that it is should.  There are MANY use cases for this, as you can see by the Apollo post that had several likes.  That way you can optimistically update mutations.  People have been doing this for years with Apollo by adding a random number to adds.  The subscription will automatically update the cache only with the changes so that the user will not get and lag time at all.  The id is replaced by the real id, and the user does not see any difference.

It would be a great speed improvement.  At least, it would create the ability to add a great speed improvement.  Not to mention you could search the cache at any given time, modify it, etc.

**Reply by kitten** [reply]
Sorry, just coming in here. I'm not sure why this sounds a bit combative, but I believe you may have proposed something that isn't quite aligning with either Apollo's cache nor Graphcache.

I'm not sure what you mean by mutations or subscriptions saving in the cache. Neither are "saved" but both "update".

Do you maybe mean that you want some kind of change when you _start_ the subscription? Other than that I can't really explain this to myself. But that being said, subscriptions in GraphQL are defined to inform you of changes but not trigger any. So any optimistic updates don't apply here because there's nothing to react to.

> I am proposing that it is should. There are MANY use cases for this, as you can see by the Apollo post that had several likes. That way you can optimistically update mutations. People have been doing this for years with Apollo by adding a random number to adds. The subscription will automatically update the cache only with the changes so that the user will not get and lag time at all.

This does sound like we aren't talking about the same thing as you are potentially 😅

Whenever a subscription event comes in then an update is made to the cached data immediately, as one would expect. So which part is optimistic here? Which post and reference are you making here, maybe if we could take a look we can get to a description of the problem here so we can find a solution for you ✌️

So as it stands, I think as @JoviDeCroock pointed out this description isn't making sense since GraphQL subscriptions and optimistic updates as the terms are defined don't make sense together 😅

**Reply by kitten** [reply]
Oh ok, wait, I think I see what you mean. Are you using subscriptions in a way in which queries are used?

So I see why you'd do that — it's convenient to operate subscriptions as "updating data" but in fact what the spec means them to be used for events which is also why "live queries" are being discussed.

The common pattern is that you start a query, you have a subscription update it, and you have mutations update it.

The idea here is that queries on the Query root type represent a tree of your server data truth, and the subscriptions and mutations describe changes. Basically, I'd say that it's a slight misuse of subscriptions (similar to how queries can technically run side effects)

**Reply by jdgamble555** [reply]
I guess to simplify my point.  The reason optimistic updates exists is for **queries** and **mutations**.  You could update the cache (whatever cache **urql** uses in the backend) by using optimistic updates.  The example is adding something to a todo list and not having to refetch the query, although this works too.

I am simply saying why can't this be done with **subscriptions**.  Obviously it doesn't work that way now.  

If it is a **slight misuse** of subscriptions, than why would optimistic updates exists at all?  We are _optimistically_ showing that the data was in fact updating in the background. 

By allowing this to be done with subscriptions, which is equally sure to happen without subscriptions, your app is faster.  I think speed alone here is the perfect use, not misuse.

Again, firebase does this within its core code.

**Reply by kitten** [reply]
You're jumping the gun here a little. So I assume I've guessed correctly as to how you're using subscriptions. However, subscriptions are a terrible way to represent server data and can't be optimistically changed because they're meant to describe updates.

Queries always represent the data tree of an application and subscriptions are limited by the spec to not take on this responsibility. This is why you can only use a single field for subscriptions for instance; that makes it very clear that they're events and not requests. Also, that's why live queries are standardised.

The typical flow of applications in Graphcache, Relay, and Apollo is to send queries and subscriptions and mutations to update this data as the application is represented by a single "state tree" from the Query root type.

We've recently updated the docs on Normalised Caching and for large parts the same principles apply to all GraphQL normalised caches. It also lays out why these patterns are the way they are https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/

However, what I think is more important here is that you're trying to force a conclusion onto us while we're trying to point out a problem with one of your premises. The premise that subscriptions may be used to represent server data state is simply incompatible with how the spec has been designed: http://spec.graphql.org/June2018/#sec-Source-Stream

If you take that on contrast to live queries, those are designed deliver updates to a query, which is a distinct difference.

Behind the scenes a normalised cache must be able to operate on a single state. If you start a subscription then there's no guarantee that a result will be available. Basically, can a cache deliver a result for a `newMessage` field? It cannot since the definition of subscriptions are event streams which are distinct from results.

The limitation here is inherent and purposeful. A subscription is defined to not have a single outcome but a stream of outcomes. If you choose to implement subscriptions by interpreting their events as a single updated outcome (which coincidentally is exactly what live queries are being defined to do) then you're subject to that limitation. For a normalised cache this could also mean that the source of truth is split. But this all would depend on the client interpreting every subscription as a single updated state, which isn't the case. This becomes more evident when considering that this limitation is also directly applied to documents for subscription operations as they're limited to a single field at a time.

Even these things considered, an optimistic result in Graphcache that's defined for a given mutation will actually update the cached data. Hence, it will update subscriptions or shared data that a subscription may also have updated. But there's no part in the specification that would allow us to synthetically create a subscription event. Instead, that's why it's common to combine a single query with one or more subscriptions. To keep the query updated via the normalised cache.

But as long as this premise isn't taken into account it would seem like your question could be solved; but rather it instead pretty much illustrates why live queries exist / are being standardised.

**Reply by jdgamble555** [reply]
Thanks @kitten for the detailed response.

So basically it is not part of the specification.  I understand that.  

> subscriptions are a terrible way to represent server data and can't be optimistically changed because they're meant to describe updates.

While I do understand why you can't change the specification, I fundamentally disagree with this statement.  I do not know the ends and outs of how the cache is intended to work, but your definition seems pretty clear.

I am perhaps suggesting a new use outside of its specification.  If you think about the user on the front end, and don't get bogged down with the specification for the cache, you can see a real use to have the DOM update before the subscription returns the results... speed.  

I believe this is why firebase chose to write their code that way.  While they don't use **graphql** (as far as I know), they do use subscriptions to keep the data up to date at all times and update the DOM first before the database gets updated on the backend and returns the data.  I believe this was a smart decision and shows why it is so fast and user friendly.

Ultimately it seems like a problem with the specification to me, as it limits the potential speed and usage of subscriptions (besides just notifications like messages etc.).

Thanks,
J

**Reply by tom2strobl** [reply]
@jdgamble555 Why you are comparing GraphQL subscriptions with Firebase over and over again is something I don't really understand if I'm afraid. Firebase has nothing to do with any GraphQL spec and we're in GraphQL userland here.
Also, Firebase is proprietary and thus in control of one hand, GraphQL isn't and therefore specs are very important to allow things like `urql` to even exist. Maybe also to add, @kitten is not talking about the the programming pattern of subscriptions but what the idiomatic term of "GraphQL subscriptions" means. 🙂

**Reply by jdgamble555** [reply]
Hi @tom2strobl - Basically because Firebase uses Web Sockets and optimistic subscription updates.  I was not aware that the specification for GraphQL Subscriptions does not allow for updating the cache / DOM first for mutations before the subscription query gets the triggered results back.  URQL and Apollo go by those specs, and I understand that now.

This whole post is to highlight the usefulness of updating the DOM (by means of the cache) for mutations before the subscription pulls the results back, which I believe is very valuable and makes apps more efficient and faster.

Again, not part of the specs, so I get that.  My argument should probably be to change [those specs](https://spec.graphql.org/June2018/)...

**Reply by kitten** [reply]
I really don't want to revive this because as I pointed out before, there are some misunderstandings here re. the spec, since GraphQL live queries are around to actively push updates from the server to the client and subscriptions push "orphan" updates from the server to the clients for entities that aren't on the `Query` schema tree outwards. There's also nothing stopping anyone from implementing mutations over Websockets which we actually support
